### DISCUSSION TITLE
Edited/Blocked Dependency updates should include a link to the branch if possible

### DISCUSSION THREAD

**Discussion by zamicly** [renovatebot/renovate#22138] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

My team recently ran into an issue where the Renovate bot that we added refused to update one of our dependencies. The Dependency Dashboard listed the update as Edited/Blocked:

![image](https://github.com/renovatebot/renovate/assets/31395342/1f59c390-a9e3-4af2-b8ed-1f3c24bc828b)

We couldn't figure out why it wouldn't update that dependency, until we realized that we had a branch laying around that was named the same as what Renovate would have named that update (`renovate/node-18.x`):

![image](https://github.com/renovatebot/renovate/assets/31395342/8e02722d-1f8f-487a-b2f4-0d5632ef4016)

For backstory, this was a branch that we manually created to fix the dependency ourselves. After we merged this change, the branch must have accidentally stuck around instead of being auto deleted. This caused Renovate to think that we were still making the change ourselves, and thus refused to auto update it itself.

After we deleted the branch, Renovate was able to correctly update that dependency.

So my proposal is two fold:
**1) Update the docs to mention this behavior**
It would be nice for it to be mentioned in the [docs](https://docs.renovatebot.com/) somewhere of this behavior. That if you have a branch named the same as what Renovate would name it, that it will block Renovate from updating that dependency. I couldn't find anywhere where this was referenced. Or for that matter, any place that described what the Edited/Blocked section actually represented.

**2) Add a link to the entry in the Edited/Blocked section that leads to the specific branch blocking that dependency from updating**
If there was a link within the Edited/Blocked section that let us know "hey this specific thing is why this dependency can't update", that would have saved us a lot of time for searching around and trying various solutions that were ultimately unsuccessful because we just had to delete that branch.

It would be great if you let me know your thoughts on this. Overall we've had tremendous success with Renovate bot for updating dependencies across all of our repos. This was the the only thorn in our side that we've found so far with it. Thank you!

**Comment by HonkingGoose** [comment]
> For backstory, this was a branch that we manually created to fix the dependency ourselves. After we merged this change, the branch must have accidentally stuck around instead of being auto deleted. This caused Renovate to think that we were still making the change ourselves, and thus refused to auto update it itself.

The Dependency Dashboard says:

> [snip] To discard all commits and start over, click on a checkbox.

So if you had clicked that checkbox, Renovate would overwrite your "custom" `renovate/node-18.x` branch. And shortly after you would get a fresh PR for the Node.js 18 update. You would not have to delete the `renovate/node-18.x` branch first. But I agree that the message needs to be easier to understand. 😉 

@rarkins want me to create an issue to work on a better text?

## Update the docs to mention this behavior

Our docs hint at the "Don't use the same branch names as Renovate"-behavior: [^1]

> ## How Renovate finds existing PRs
> 
> Renovate does not need to maintain any database/state about open or closed Pull Requests.
> Instead, it uses the code platform's APIs to search and find such PRs.
> 
> Renovate finds existing PRs (open or closed) by matching both:
> 
> - the branch name, for example: `renovate/lodash-4.x`,
> - _and_ the Pull Request title, for example: `Update lodash to v4.17.21`
> 
> In cases like the above, there is typically one existing PR with a matching branch name and PR title.
> But if you group PRs and use titles like "All non-major updates", then multiple past PRs may match.

@rarkins do you want me to add an admonition warning like this one, to better explain this gotcha?

> **Warning**
> Avoid the `renovate/` prefix for your own branches. This prevents conflicts between your branch and Renovate's update branches."

## Describe what the Edited/Blocked section represents

> Or for that matter, any place that described what the Edited/Blocked section actually represented.

You're right, I can't find this in our docs. Possible solutions:

1. Explain in Dependency Dashboard issue body text what each section means
2. Explain in docs (new page) what each Dependency Dashboard section means
3. Explain in Dependency Dashboard issue _and_ in the docs

I'll let @rarkins pick the option, and if they want an issue from me with their option.

## Link to edited/blocked branch from Dashboard

This sounds like a nice idea. I'll let @rarkins decide if they want an issue for this idea as well. 🙈 

[^1]: [Renovate docs, how Renovate finds existing PRs](https://docs.renovatebot.com/key-concepts/pull-requests/#how-renovate-finds-existing-prs)

**Reply by rarkins** [reply]
We don't want to make the Dependency Dashboard text much longer, although if it can be made more explanatory within the same type of space the it's fine. Realistically I doubt any documentation would have prevented the user pushing to a `renovate/` branch in the first place.

Otherwise it would be good too explain the sections of Dependency Dashboard from within this page: https://docs.renovatebot.com/key-concepts/dashboard/

**Reply by HonkingGoose** [reply]
> Realistically I doubt any documentation would have prevented the user pushing to a `renovate/` branch in the first place.

I agree.

> Otherwise it would be good too explain the sections of Dependency Dashboard from within this page: https://docs.renovatebot.com/key-concepts/dashboard/

I like that idea! I'll make an issue for it, and assign myself:

- #22234

**Reply by zamicly** [reply]
Thanks for the quick response on this!

> So if you had clicked that checkbox, Renovate would overwrite your "custom" renovate/node-18.x branch. And shortly after you would get a fresh PR for the Node.js 18 update.

@HonkingGoose So we actually did try clicking this checkbox, but when renovate ran, he would just uncheck that checkbox in the dashboard without actually making any changes to the branch. I'm not sure if this is expected behavior? Did I miss something else?

> Our docs hint at the "Don't use the same branch names as Renovate"-behavior

@HonkingGoose ah gotcha gotcha. It's up to you all how you want to handle it obviously, but a warning or something like that would be nice imo ¯\\\_(ツ)_/¯ 

> I like that idea! I'll make an issue for it, and assign myself

@HonkingGoose @rarkins Thank you both for the input on this discussion!

**Reply by HonkingGoose** [reply]
As far as I know Renovate should have overwritten that branch after you selected the checkbox... @rarkins do you have any idea why Renovate would keep the branch as-is?

**Reply by rarkins** [reply]
No, it would require logs/reproduction to look into

**Reply by zamicly** [reply]
Gotcha, I can attempt to break it again (now that I know what actually breaks it), then report back anything I find. All we did was have that branch laying around. It was outdated, so not sure if that's what caused it maybe?

Regarding the logs, we didn't see anything helpful in them when renovate ran:
```
 INFO: Repository started (repository=<REDACTED>)
       "renovateVersion": "34.153.1"
 INFO: Dependency extraction complete (repository=<REDACTED>, baseBranch=main)
       "stats": {
         "managers": {
           "dockerfile": {"fileCount": 1, "depCount": 1},
           "npm": {"fileCount": 2, "depCount": 55}
         },
         "total": {"fileCount": 3, "depCount": 56}
       }
 INFO: Repository finished (repository=<REDACTED>)
       "cloned": true,
       "durationMs": 11369 
```

**Reply by rarkins** [reply]
You need debug logs

**Reply by zamicly** [reply]
So I haven't got approval to share our full logs yet, but if you wanted to recreated it, it's pretty simple too. The process to recreate is:
	1.	Name a branch the same as what renovate would name it
	2.	Open and merge a PR for that branch
	3.	Either don't cleanup/delete that branch or create a new branch with the same name
	4.	Renovate blocks any further updates to that dependency automatically and adds that update to the Edited/Blocked section in the Dependency Dashboard
	5.	If you click that checkbox on the dashboard
	6.     It refuses to actually update that branch

In reviewing the debug logs, though, it looks like the reason it refuses to update is that dependency is because it sees the branch and it sees the previously closed pr. But it errors out because the SHA between the 2 is different.

A couple brief logs that I can share are:
```
  {<REDACTED>,"branch":"renovate/node-18.x","oldPrNumber":179,"oldPrSha":"fba60b597b1acdcbb63843739f79d9d3c8134d2a","branchSha":"bda4094c71a78b7e77ecbe411ae902fe967f6386","msg":"Found old PR but the SHA is different","time":"2023-05-25T12:17:39.779Z","v":0}
```

and

```
  {<REDACTED>,"baseBranch":"main","branch":"renovate/node-18.x","err":{"task":{"commands":["merge","--abort"],"format":"utf-8","parser":"[function]"},"message":"fatal: There is no merge to abort (MERGE_HEAD missing).\n","stack":"Error: fatal: There is no merge to abort (MERGE_HEAD missing).\n\n    at Object.action (/opt/buildpack/tools/renovate/34.153.1/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/opt/buildpack/tools/renovate/34.153.1/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /opt/buildpack/tools/renovate/34.153.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/opt/buildpack/tools/renovate/34.153.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/opt/buildpack/tools/renovate/34.153.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/opt/buildpack/tools/renovate/34.153.1/node_modules/simple-git/dist/cjs/index.js:55:24)"},"msg":"isBranchConflicted: cleanup error","time":"2023-05-25T12:17:40.152Z","v":0}
```

Not sure if any of this helps
